Add patch to prevent NPE of pos

liach opened the pull request:
With a modded way to sleep, e.x. a minecart that allows sleeping, the bedPosition may be null and cause a npe in block pos boundary check.

Signed-off-by: liach <liach@users.noreply.github.com>

Edit: I really wish people can review this pull so that I get this *crash* fixed. It is a crash.

Edit 2: Why would I fix the crash instead of setting a position? Because that position is used to reset spawn if it is not set to null.

liach committed:
Add patch to prevent NPE of pos

With a modded way to sleep, e.x. a minecart that allows sleeping, the bedPosition may be null and cause a npe in block pos boundary check.

Signed-off-by: liach <liach@users.noreply.github.com>

liach commented:
It's very weird that I discovered an issue with entity spawning with incorrect factory on the client side after getting an entity message packet. I will make a change for that too so that my test mod works correctly.

liach committed:
Adds a sleeping minecart test

Also fixed an issue with entity spawn message not using registry factories

Signed-off-by: liach <liach@users.noreply.github.com>

liach commented on src/test/java/net/minecraftforge/debug/entity/player/PlayerSleepWithoutPosTest.java:
```diff
@@ -0,0 +1,340 @@
+@Mod(modid = MOD_ID, name = "Player Sleep Without Position Test", acceptableRemoteVersions = "*")
+public final class PlayerSleepWithoutPosTest
+{
+    private static final boolean ENABLED = true;
```
I will edit this field later.

liach commented on src/test/resources/assets/player_sleep_without_pos_test/lang/en_US.lang:
```diff
@@ -0,0 +1,3 @@
+player_sleep_without_pos_test.key.sleep=Minecart Sleep Key
```
I will edit the file name later.

liach committed:
Rename the language file and disable test mod by default

The language file does not work because in dev env only classes are loaded; no assets is loaded.

Signed-off-by: liach <liach@users.noreply.github.com>

HenryLoenwind commented:
You may want to remove the `cls = er.getEntityClass()`, too. The newInstance() is the only consumer.

And go though each use of `net.minecraftforge.fml.common.registry.EntityRegistry.EntityRegistration.entityClass` to see if that's wrong, too. (only `net.minecraftforge.fml.common.registry.EntityRegistry.addSpawn()` and `removeSpawn()` as far as I can see.)

bs2609 commented:
Related: #4845, #4846

liach committed:
Refactored entity registration a bit per suggestion

Should be backward compatible

Now entity registration require an entity entry to be present in the registry.

Signed-off-by: liach <liach@users.noreply.github.com>

liach commented:
@HenryLoenwind that was in fact the only place where that class was used.

For both henry and @bs2609 , would you comment on my approach of entity registration changes?

HenryLoenwind commented:
I'm no expert on entities, but that patch looks good to me.

You may want to split off your original NPE patch from this to get that one merged sooner. ;)

liach commented:
uhh, i opened my pr for the npe patch and now rediscovered the old bug.
